<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="John Blischak" />

<meta name="date" content="2017-08-23" />

<title>Identifying single cells</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">singlecell-qtl</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="contributing.html">Contributing</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/singlecell-qtl">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Identifying single cells</h1>
<h4 class="author"><em>John Blischak</em></h4>
<h4 class="date"><em>2017-08-23</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<p><strong>Last updated:</strong> 2017-10-24</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> cec0e85</p>
<p><a href="https://catalog.coriell.org/0/Sections/Collections/NHGRI/1000IndSamples.aspx?PgId=665&amp;did=MGENOME27">Link to Yoruba samples on Coriell</a></p>
<div id="setup" class="section level2">
<h2>Setup</h2>
<pre class="r"><code>library(&quot;cowplot&quot;)
library(&quot;dplyr&quot;)
library(&quot;ggplot2&quot;)
library(&quot;stringr&quot;)
library(&quot;tidyr&quot;)
theme_set(theme_cowplot())
library(&quot;Biobase&quot;)</code></pre>
<p>Import data.</p>
<pre class="r"><code>fname &lt;- Sys.glob(&quot;../data/eset/*.rds&quot;)
eset &lt;- Reduce(combine, Map(readRDS, fname))
eset</code></pre>
<pre><code>ExpressionSet (storageMode: lockedEnvironment)
assayData: 54792 features, 960 samples 
  element names: exprs 
protocolData: none
phenoData
  sampleNames: 03162017-A01 03162017-A02 ... 04202017-H12 (960
    total)
  varLabels: experiment well ... valid_id (39 total)
  varMetadata: labelDescription
featureData
  featureNames: ENSG00000000003 ENSG00000000005 ... WBGene00235374
    (54792 total)
  fvarLabels: chr start ... source (6 total)
  fvarMetadata: labelDescription
experimentData: use &#39;experimentData(object)&#39;
Annotation:  </code></pre>
</div>
<div id="how-many-of-the-identities-are-plausible" class="section level2">
<h2>How many of the identities are plausible?</h2>
<pre class="r"><code>table(eset$valid_id)</code></pre>
<pre><code>
FALSE  TRUE 
   98   862 </code></pre>
<pre class="r"><code>mean(eset$valid_id)</code></pre>
<pre><code>[1] 0.8979167</code></pre>
</div>
<div id="plot-per-c1-chip" class="section level2">
<h2>Plot per C1 chip</h2>
<pre class="r"><code>anno &lt;- pData(eset)
df_e &lt;- anno %&gt;%
  group_by(experiment) %&gt;%
  summarize(i1 = unique(individual.1),
            i2 = unique(individual.2),
            i3 = unique(individual.3),
            i4 = unique(individual.4))
op &lt;- par(cex = 2, las = 2, mfrow = c(2, 5))
for (i in 1:nrow(df_e)) {
  e &lt;- df_e$experiment[i]
  n1 &lt;- sum(anno$chip_id[anno$experiment == e] == df_e$i1[i])
  n2 &lt;- sum(anno$chip_id[anno$experiment == e] == df_e$i2[i])
  n3 &lt;- sum(anno$chip_id[anno$experiment == e] == df_e$i3[i])
  n4 &lt;- sum(anno$chip_id[anno$experiment == e] == df_e$i4[i])
  n_other &lt;- 96 - n1 - n2 -n3 - n4
  barplot(c(n1, n2, n3, n4, n_other), main = sprintf(&quot;C1 chip %s&quot;, e),
          names.arg = c(df_e$i1[i], df_e$i2[i], df_e$i3[i],df_e$i4[i], &quot;Other&quot;),
          ylab = &quot;Number of single cells&quot;)
}</code></pre>
<p><img src="figure/identity.Rmd/ind-per-chip-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>par(op)</code></pre>
<p>Colored by individual. The most worrisome are the last three C1 chips (04132017, 04142017, 04202017), which all have high numbers of assignments coming from one or more individuals that were not on that chip.</p>
<pre class="r"><code>anno$valid_id_display &lt;- factor(anno$valid_id, levels = c(TRUE, FALSE),
                                labels = c(&quot;Valid&quot;, &quot;Invalid&quot;))
ggplot(anno, aes(x = valid_id_display, fill = chip_id)) +
  geom_bar() +
  facet_wrap(~experiment, nrow = 2) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        legend.position = &quot;bottom&quot;) +
  labs(x = &quot;&quot;, y = &quot;Number of single cells&quot;,
       title = &quot;Valid vs. invalid assignments colored by individual&quot;)</code></pre>
<p><img src="figure/identity.Rmd/ind-per-chip-color-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="expected-versus-observed-cells" class="section level2">
<h2>Expected versus observed cells</h2>
<p>Each individual is expected to get 25% of the wells (i.e.Â 24) for each C1 chip it is on.</p>
<pre class="r"><code>expected &lt;- df_e %&gt;% select(starts_with(&quot;i&quot;)) %&gt;% unlist %&gt;% table * 96 / 4
expected &lt;- as.data.frame(expected)
colnames(expected) &lt;- c(&quot;ind&quot;, &quot;count&quot;)
observed &lt;- table(anno$chip_id)
observed &lt;- as.data.frame(observed)
colnames(observed) &lt;- c(&quot;ind&quot;, &quot;count&quot;)
df_obs &lt;- merge(expected, observed, by = &quot;ind&quot;, suffixes = c(&quot;.exp&quot;, &quot;.obs&quot;),
                all.x = TRUE, sort = TRUE)
# Some individuals were assigned zero cells
df_obs$count.obs[is.na(df_obs$count.obs)] &lt;- 0
df_obs</code></pre>
<pre><code>       ind count.exp count.obs
1  NA18498        72        21
2  NA18502       144       219
3  NA18505        96        60
4  NA18508        24        25
5  NA18510        24         0
6  NA18520       168       163
7  NA18852       144       104
8  NA18856        72        88
9  NA19092        48        53
10 NA19098        96        86
11 NA19190        24        19
12 NA19203        48        29</code></pre>
<pre class="r"><code># The max observed number of cells for any indvidual
max_cells &lt;- max(df_obs$count.exp, df_obs$count.obs)
ggplot(df_obs, aes(x = count.exp, y = count.obs)) +
  geom_text(aes(label = ind)) +
  geom_abline(intercept = 0, slope = 1, color = &quot;red&quot;, linetype = &quot;dashed&quot;) +
  xlim(0, max_cells) + ylim(0, max_cells) +
  labs(x = &quot;Expected number of cells per individual&quot;,
       y = &quot;Observed number of cells per individual&quot;,
       title = &quot;Observed versus expected number of cells&quot;)</code></pre>
<p><img src="figure/identity.Rmd/obs-v-exp-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="unexpectedly-high-counts" class="section level2">
<h2>Unexpectedly high counts</h2>
<pre class="r"><code>observed %&gt;% filter(count &gt; 10, !(ind %in% expected$ind)) %&gt;%
  arrange(desc(count))</code></pre>
<pre><code>      ind count
1 NA18862    49
2 NA18507    23</code></pre>
</div>
<div id="potential-reasons-for-failure" class="section level2">
<h2>Potential reasons for failure</h2>
<p>Sequencing depth</p>
<pre class="r"><code>reason_raw &lt;- ggplot(anno, aes(x = valid_id, y = raw)) +
  geom_boxplot() +
  scale_x_discrete(labels = c(&quot;Wrong&quot;, &quot;Correct&quot;)) +
  labs(x = &quot;Predicted individual of origin&quot;,
       y = &quot;Raw number of reads&quot;,
       title = &quot;Raw number of reads&quot;)
reason_raw</code></pre>
<p><img src="figure/identity.Rmd/reason-raw-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Percent mapped</p>
<pre class="r"><code>reason_raw %+% aes(y = mapped / raw * 100) +
  labs(y = &quot;Percent of reads mapped&quot;,
       title = &quot;Percent of reads mapped&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-mapped-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Cell number</p>
<pre class="r"><code>barplot(with(anno, table(valid_id, cell_number)), beside = TRUE,
        col = c(&quot;red&quot;, &quot;black&quot;),
        xlab = &quot;Number of cells in well&quot;, main = &quot;Number of cells in well&quot;,
        ylab = &quot;Number of cells correct (black) versus wrong (red)&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-cell-number-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Concentration</p>
<pre class="r"><code>reason_raw %+% aes(y = concentration) +
  labs(y = &quot;Concentration&quot;,
       title = &quot;Concentration&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-concentation-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>TRA-1-60</p>
<pre class="r"><code>barplot(with(anno, table(valid_id, tra1.60)), beside = TRUE,
        col = c(&quot;red&quot;, &quot;black&quot;),
        xlab = &quot;TRA-1-60&quot;, main = &quot;TRA-1-60&quot;,
        ylab = &quot;Number of cells correct (black) versus wrong (red)&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-tra-1-60-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>SNPs with at least one overlapping read</p>
<pre class="r"><code>reason_raw %+% aes(y = snps_w_min) +
  labs(y = &quot;SNPs with at least one overlapping read&quot;,
       title = &quot;SNPs with at least one overlapping read&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-snps-w-min-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>chipmix</p>
<pre class="r"><code>reason_raw %+% aes(y = chipmix) +
  labs(y = &quot;chipmix&quot;, title = &quot;chipmix&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-chipmix-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>freemix</p>
<pre class="r"><code>reason_raw %+% aes(y = freemix) +
  labs(y = &quot;freemix&quot;, title = &quot;freemix&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-freemix-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>chipmix versus freemix</p>
<pre class="r"><code>ggplot(anno, aes(x = freemix, y = chipmix, color = valid_id)) +
  geom_point() +
  facet_wrap(~valid_id) +
  labs(title = &quot;chipmix versus freemix&quot;)</code></pre>
<p><img src="figure/identity.Rmd/chipmix-v-freemix-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>chipmix versus SNPs with overlapping reads</p>
<pre class="r"><code>ggplot(anno, aes(x = snps_w_min, y = chipmix, color = valid_id)) +
  geom_point() +
  facet_wrap(~valid_id) +
  labs(title = &quot;chipmix versus SNPs with at least one overlapping read&quot;,
       x = &quot;SNPs with at least one overlapping read&quot;)</code></pre>
<p><img src="figure/identity.Rmd/chipmix-v-snps-w-min-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>chipmix versus raw number of reads</p>
<pre class="r"><code>ggplot(anno, aes(x = raw, y = chipmix, color = valid_id)) +
  geom_point() +
  facet_wrap(~valid_id) +
  labs(title = &quot;chipmix versus number of raw reads&quot;,
       x = &quot;Number of raw reads&quot;)</code></pre>
<p><img src="figure/identity.Rmd/chipmix-v-raw-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<!-- Insert the session information into the document -->
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.3.2 (2016-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.2 (Nitrogen)

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  methods   stats     graphics  grDevices utils     datasets 
[8] base     

other attached packages:
[1] Biobase_2.34.0      BiocGenerics_0.20.0 tidyr_0.6.1        
[4] stringr_1.2.0       dplyr_0.5.0         cowplot_0.7.0      
[7] ggplot2_2.2.1      

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.10     knitr_1.16       magrittr_1.5     munsell_0.4.3   
 [5] colorspace_1.3-2 R6_2.2.0         plyr_1.8.4       tools_3.3.2     
 [9] grid_3.3.2       gtable_0.2.0     DBI_0.6-1        git2r_0.19.0    
[13] htmltools_0.3.6  yaml_2.1.13      lazyeval_0.2.0   rprojroot_1.2   
[17] digest_0.6.12    assertthat_0.1   tibble_1.2       evaluate_0.10   
[21] rmarkdown_1.6    labeling_0.3     stringi_1.1.2    scales_0.4.1    
[25] backports_1.0.5 </code></pre>
</div>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
