<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2017-12-15 Fri 19:20 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Power to detect QTLs in single cell data</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">Power to detect QTLs in single cell data</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orgf9fdb5e">Introduction</a></li>
<li><a href="#org2a3b9fe">Resample scRNA-Seq</a></li>
<li><a href="#org0b73441">Plot concordance between simulated and real data</a></li>
<li><a href="#orgab2a8e9">Simulate at a bulk RNA-Seq QTL</a></li>
<li><a href="#org7eb4bf1">Open problems</a></li>
</ul>
</div>
</div>

<div id="outline-container-orgf9fdb5e" class="outline-2">
<h2 id="orgf9fdb5e">Introduction</h2>
<div class="outline-text-2" id="text-orgf9fdb5e">
<p>
Our goal here is estimate power to detect mean and variance-effect QTLs. The
strategy is to resample the single cell RNA-Seq data to generate datasets
with equal patterns of dropout, overdispersion, mean-variance
dependence, etc.
</p>
</div>
</div>

<div id="outline-container-org2a3b9fe" class="outline-2">
<h2 id="org2a3b9fe">Resample scRNA-Seq</h2>
<div class="outline-text-2" id="text-org2a3b9fe">
<p>
Assuming the true single cells all came from one individual, generate new
single cells by sampling counts for each gene iid. with replacement. This
method assumes genes are independent, which is justified only if we perform
analysis one gene at a time.
</p>

<p>
To generate a new data set (with multiple individuals), randomly sample
individuals with replacement, and generate a new set of single cells for each
sampled individual.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="orgf37f61e"><span class="org-keyword">def</span> <span class="org-function-name">sample_cells</span>(umi, n=1):
  <span class="org-doc">"""Return a sampled matrix of UMI counts</span>

<span class="org-doc">  Assume the input matrix is UMI for a single individal. Sample gene counts</span>
<span class="org-doc">  i.i.d. from the empirical distribution of counts</span>

<span class="org-doc">  """</span>
  <span class="org-variable-name">result</span> = np.zeros((umi.shape[0], n))
  <span class="org-keyword">for</span> i, row <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(umi.values):
    <span class="org-variable-name">result</span>[i] = np.random.choice(row, size=n, replace=<span class="org-constant">True</span>)
  <span class="org-keyword">return</span> pd.DataFrame(result, index=umi.index)

<span class="org-keyword">def</span> <span class="org-function-name">sample_individuals</span>(umi, annotations, num_individuals=<span class="org-constant">None</span>, num_cells=100):
  <span class="org-doc">"""Return a sampled matrix of UMI counts for multiple individuals</span>

<span class="org-doc">  num_individuals - # individuals to sample (default: equal number)</span>
<span class="org-doc">  num_cells - # cells to sample for each sampled individual</span>

<span class="org-doc">  Returns:</span>

<span class="org-doc">  umi - matrix of counts</span>
<span class="org-doc">  individuals - list of resampled IDs</span>

<span class="org-doc">  """</span>
  <span class="org-keyword">if</span> num_individuals <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-variable-name">num_individuals</span> = <span class="org-builtin">len</span>(lookup)
  <span class="org-variable-name">individuals</span> = <span class="org-builtin">list</span>(np.random.choice(annotations[<span class="org-string">'chip_id'</span>].unique(), size=num_individuals, replace=<span class="org-constant">True</span>))
  <span class="org-variable-name">result</span> = pd.DataFrame(index=umi.index)
  <span class="org-keyword">for</span> k <span class="org-keyword">in</span> individuals:
    <span class="org-variable-name">result</span> = result.merge(
      sample_cells(umi.loc[:,(annotations[<span class="org-string">'chip_id'</span>] == k).values], n=num_cells),
      left_index=<span class="org-constant">True</span>,
      right_index=<span class="org-constant">True</span>)
  <span class="org-keyword">return</span> result, individuals
</pre>
</div>
</div>
</div>

<div id="outline-container-org0b73441" class="outline-2">
<h2 id="org0b73441">Plot concordance between simulated and real data</h2>
<div class="outline-text-2" id="text-org0b73441">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">pools</span> = [cpm(sample_cells(umi.loc[:,(annotations[<span class="org-string">'chip_id'</span>] == <span class="org-string">'NA18507'</span>).values], n=100).agg(np.<span class="org-builtin">sum</span>, axis=1), log2=<span class="org-constant">True</span>).to_frame()
         <span class="org-keyword">for</span> _ <span class="org-keyword">in</span> <span class="org-builtin">range</span>(2)]

plot_concordance(
  x=pools[0],
  xlabel=<span class="org-string">'Simulated single cell $\log_2(CPM + 1)$'</span>,
  y=bulk[<span class="org-string">'NA18507'</span>].to_frame(),
  ylabel=<span class="org-string">'Bulk $\log_2(TPM + 1)$'</span>,
  title=<span class="org-string">'NA18507'</span>,
  filename=<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/analysis/figure/power.org/sim/sc-vs-bulk.png'</span>)
plot_concordance(
  x=pools[0],
  xlabel=<span class="org-string">'Simulated single cell $\log_2(CPM + 1)$'</span>,
  y=pools[1],
  ylabel=<span class="org-string">'Simulated single cell $\log_2(CPM + 1)$'</span>,
  title=<span class="org-string">'NA18507'</span>,
  filename=<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/analysis/figure/power.org/sim/sc-vs-sc.png'</span>,
  gridsize=15)
</pre>
</div>

<p>
The true concordance, <a href="sc-vs-bulk.html#orgf1f819a">as previously computed</a>:
</p>

<p>
<img src="figure/sc-vs-bulk.org/normed/NA18507.png" alt="NA18507.png">
<img src="figure/sc-vs-bulk.org/subsets/NA18507-100.png" alt="NA18507-100.png">
</p>

<p>
Generated single cell. vs bulk and single cell vs. single cell concordance:
</p>

<p>
<img src="figure/power.org/sim/sc-vs-bulk.png" alt="sc-vs-bulk.png">
<img src="figure/power.org/sim/sc-vs-sc.png" alt="sc-vs-sc.png">
</p>
</div>
</div>

<div id="outline-container-orgab2a8e9" class="outline-2">
<h2 id="orgab2a8e9">Simulate at a bulk RNA-Seq QTL</h2>
<div class="outline-text-2" id="text-orgab2a8e9">
<p>
Consider the gene <i>ZWIM7</i>, for which the strongest mean eQTL is
<code>rs73276049</code>. We previously plotted the distribution of UMI counts stratified
by genotype:
</p>


<div class="figure">
<p><img src="figure/sc-stratified-geno.org/214941.png" alt="214941.png">
</p>
</div>

<p>
Get the genotype for each individual:
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">extract_genotypes</span>(f, rsid):
  <span class="org-variable-name">head</span> = <span class="org-builtin">next</span>(f).split()
  <span class="org-keyword">for</span> line <span class="org-keyword">in</span> f:
    <span class="org-variable-name">record</span> = line.split()
    <span class="org-keyword">if</span> record[2] == rsid:
      <span class="org-keyword">return</span> {<span class="org-string">'NA{}'</span>.<span class="org-builtin">format</span>(k): <span class="org-builtin">int</span>(np.<span class="org-builtin">round</span>(<span class="org-builtin">float</span>(v))) <span class="org-keyword">for</span> k, v <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(head, record) <span class="org-keyword">if</span> k[0] == <span class="org-string">'1'</span>}
  <span class="org-keyword">return</span> <span class="org-builtin">dict</span>()
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">with</span> gzip.<span class="org-builtin">open</span>(<span class="org-string">'/project/compbio/jointLCLs/genotype/hg19/YRI/genotypesYRI.gen.txt.gz'</span>, <span class="org-string">'rt'</span>) <span class="org-keyword">as</span> f:
  <span class="org-variable-name">lookup</span> = extract_genotypes(f, <span class="org-string">'rs73276049'</span>)
</pre>
</div>

<p>
Simulate new cohorts varying the sample size, stratify on genotype, and plot:
</p>

<ol class="org-ol">
<li>the distribution of simulated UMI counts</li>
<li>the mean UMI count</li>
</ol>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">colors</span> = {k: matplotlib.cm.get_cmap(<span class="org-string">'Set1'</span>).colors[k] <span class="org-keyword">for</span> k <span class="org-keyword">in</span> <span class="org-builtin">range</span>(-1, 3)}
<span class="org-keyword">for</span> n <span class="org-keyword">in</span> [20, 40, 60, 80]:
  <span class="org-variable-name">sampled_umi</span>, <span class="org-variable-name">individuals</span> = sample_individuals(umi, annotations, num_individuals=n)
  plt.clf()
  <span class="org-comment-delimiter"># </span><span class="org-comment">We need to handle missing genotypes because NA19092 isn't in the genotype</span>
  <span class="org-comment-delimiter"># </span><span class="org-comment">matrix</span>
  <span class="org-keyword">for</span> k, g <span class="org-keyword">in</span> sampled_umi.loc[<span class="org-string">'ENSG00000214941'</span>].groupby([lookup.get(k, -1) <span class="org-keyword">for</span> k <span class="org-keyword">in</span> individuals <span class="org-keyword">for</span> _ <span class="org-keyword">in</span> <span class="org-builtin">range</span>(100)]):
    <span class="org-keyword">if</span> k &gt;= 0:
      plt.hist(g, histtype=<span class="org-string">'step'</span>, bins=np.arange(g.<span class="org-builtin">max</span>()).astype(np.float32), color=colors[k])
      plt.axvline(x=g.mean(), color=colors[k], label=<span class="org-builtin">int</span>(k))
  plt.title(<span class="org-string">'ZWIM7, {} individuals, 100 cells'</span>.<span class="org-builtin">format</span>(n))
  plt.ylabel(<span class="org-string">'Number of cells'</span>)
  plt.xlabel(<span class="org-string">'UMI count'</span>)
  plt.gca().legend()
  plt.savefig(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/analysis/figure/power.org/sim/ZWIM7-{}.png'</span>.<span class="org-builtin">format</span>(n))
</pre>
</div>

<p>
<img src="figure/power.org/sim/ZWIM7-20.png" alt="ZWIM7-20.png">
<img src="figure/power.org/sim/ZWIM7-40.png" alt="ZWIM7-40.png">
<img src="figure/power.org/sim/ZWIM7-60.png" alt="ZWIM7-60.png">
<img src="figure/power.org/sim/ZWIM7-80.png" alt="ZWIM7-80.png">
</p>
</div>
</div>

<div id="outline-container-org7eb4bf1" class="outline-2">
<h2 id="org7eb4bf1">Open problems</h2>
<div class="outline-text-2" id="text-org7eb4bf1">
<ol class="org-ol">
<li>If we estimate the effect size distribution from real data, and use the
estimated effect sizes to perform the power calculation, we need to
account for winner's curse</li>
<li>Confounder correction on the UMI counts</li>
<li>Confounder correction on the genotypes</li>
</ol>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2017-12-15 Fri 19:20</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
