<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-03-28 Wed 20:24 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Single cell/bulk RNA-Seq concordance</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Single cell/bulk RNA-Seq concordance</h1>

<div id="outline-container-org7e47893" class="outline-2">
<h2 id="org7e47893">Introduction</h2>
<div class="outline-text-2" id="text-org7e47893">
<p>
Here, we investigate the correlation between single cell RNA-Seq and bulk
RNA-Seq on samples from the same cell type in the same individuals. Our goal
is to qualitatively assess our ability to call QTLs from scRNA-Seq.
</p>
</div>
</div>

<div id="outline-container-orge2dd258" class="outline-2">
<h2 id="orge2dd258">Read the data</h2>
<div class="outline-text-2" id="text-orge2dd258">
<p>
Read and QC the single cell counts.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">umi</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-counts.txt.gz'</span>, index_col=0)
<span class="org-variable-name">annotations</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-annotation.txt'</span>)
<span class="org-variable-name">keep_samples</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/quality-single-cells.txt'</span>, index_col=0, header=<span class="org-constant">None</span>)
<span class="org-variable-name">keep_genes</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/genes-pass-filter.txt'</span>, index_col=0, header=<span class="org-constant">None</span>)
<span class="org-variable-name">umi</span> = umi.loc[keep_genes.values.ravel(),keep_samples.values.ravel()]
<span class="org-variable-name">annotations</span> = annotations.loc[keep_samples.values.ravel()]
</pre>
</div>

<p>
Our preliminary investigation into concordance between single cell log CPM
and bulk log RPKM did not give good results. Therefore, we <a href="kallisto.html">re-processed the
iPSC bulk RNA-Seq</a> using <code>kallisto</code>. Read the TPM matrix.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">bulk_log_tpm</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/kallisto/bulk-ipsc-tpm.txt.gz'</span>, header=<span class="org-constant">None</span>, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">bulk_log_tpm</span> = np.log(bulk_log_tpm.pivot(columns=0, index=1, values=2) + 1) / np.log(2)
<span class="org-variable-name">bulk_log_tpm.index</span> = [x.split(<span class="org-string">'.'</span>)[0] <span class="org-keyword">for</span> x <span class="org-keyword">in</span> bulk_log_tpm.index]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="org80f2327"><span class="org-keyword">def</span> <span class="org-function-name">plot_concordance</span>(x, y, title, filename, xlabel=<span class="org-constant">None</span>, ylabel=<span class="org-constant">None</span>, **kwargs):
  <span class="org-doc">"""Plot hexbin of concordance"""</span>
  <span class="org-variable-name">merged</span> = x.merge(y, left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
  <span class="org-variable-name">merged.columns</span> = [<span class="org-string">'x'</span>, <span class="org-string">'y'</span>]
  <span class="org-variable-name">lim</span> = [merged.<span class="org-builtin">min</span>().<span class="org-builtin">min</span>(), merged.<span class="org-builtin">max</span>().<span class="org-builtin">max</span>()]
  plt.clf()
  <span class="org-keyword">if</span> <span class="org-string">'gridsize'</span> <span class="org-keyword">not</span> <span class="org-keyword">in</span> kwargs:
    <span class="org-variable-name">kwargs</span>[<span class="org-string">'gridsize'</span>] = 40
  plt.hexbin(merged[<span class="org-string">'x'</span>], merged[<span class="org-string">'y'</span>], cmap=colorcet.cm[<span class="org-string">'blues'</span>], **kwargs)
  <span class="org-variable-name">cb</span> = plt.colorbar()
  cb.set_label(<span class="org-string">'Number of genes'</span>)
  plt.plot(lim, lim, color=<span class="org-string">'red'</span>)
  <span class="org-variable-name">ax</span> = plt.gca()
  ax.set_xlim([merged[<span class="org-string">'x'</span>].<span class="org-builtin">min</span>(), merged[<span class="org-string">'x'</span>].<span class="org-builtin">max</span>()])
  ax.set_ylim([merged[<span class="org-string">'y'</span>].<span class="org-builtin">min</span>(), merged[<span class="org-string">'y'</span>].<span class="org-builtin">max</span>()])
  plt.title(title)
  <span class="org-keyword">if</span> xlabel <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-variable-name">xlabel</span> = <span class="org-string">'scRNA-Seq $\log_2(\mathrm{CPM} + 1) $'</span>
  <span class="org-keyword">if</span> ylabel <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-variable-name">ylabel</span> = <span class="org-string">'Bulk RNA-Seq $\log_2(\mathrm{TPM} + 1)$'</span>
  plt.xlabel(xlabel)
  plt.ylabel(ylabel)
  plt.savefig(filename)

<span class="org-keyword">def</span> <span class="org-function-name">cpm</span>(counts, size=<span class="org-constant">None</span>, log2=<span class="org-constant">False</span>):
  <span class="org-keyword">if</span> size <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-variable-name">size</span> = counts.<span class="org-builtin">sum</span>(axis=0)
  <span class="org-variable-name">cpm</span> = counts / size * 1e6
  <span class="org-keyword">if</span> log2:
    <span class="org-variable-name">cpm</span> = np.log(cpm + 1) / np.log(2)
  <span class="org-keyword">return</span> cpm
</pre>
</div>
</div>
</div>

<div id="outline-container-orgc62a880" class="outline-2">
<h2 id="orgc62a880">Plot bulk vs. pooled single cells</h2>
<div class="outline-text-2" id="text-orgc62a880">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">plot_concordance_by_individual</span>(umi, annotations, bulk, output_dir):
  <span class="org-variable-name">bulk</span>, <span class="org-variable-name">pooled_cpm</span> = bulk.align(
    cpm(umi.groupby(by=annotations[<span class="org-string">'chip_id'</span>].values, axis=1).agg(np.<span class="org-builtin">sum</span>),
        size=annotations.groupby(<span class="org-string">'chip_id'</span>)[<span class="org-string">'mol_hs'</span>].agg(np.<span class="org-builtin">sum</span>), log2=<span class="org-constant">True</span>),
    axis=1, join=<span class="org-string">'inner'</span>)
  <span class="org-keyword">for</span> k <span class="org-keyword">in</span> bulk:
    plot_concordance(
      x=pooled_cpm[k].to_frame(),
      y=bulk[k].to_frame(),
      title=k,
      filename=<span class="org-string">'{}/{}.svg'</span>.<span class="org-builtin">format</span>(output_dir, k))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plot_concordance_by_individual(
  umi,
  annotations,
  bulk_log_tpm,
  <span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/analysis/figure/sc-vs-bulk.org/normed'</span>)
</pre>
</div>

<p>
Plot several examples.
</p>

<div class="org-src-container">
<pre class="src src-ipython">np.random.seed(0)
<span class="org-keyword">for</span> k <span class="org-keyword">in</span> np.random.choice(annotations[<span class="org-string">'chip_id'</span>].unique(), 4):
  <span class="org-keyword">print</span>(<span class="org-string">'file:figure/sc-vs-bulk.org/normed/{}.svg'</span>.<span class="org-builtin">format</span>(k))
</pre>
</div>

<p>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/normed/NA19130.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/normed/NA19102.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/normed/NA19143.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/normed/NA19098.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>

<p>
Regress log CPM on log TPM per individual.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">log_cpm</span> = cpm(umi.groupby(by=annotations[<span class="org-string">'chip_id'</span>].values, axis=1).agg(np.<span class="org-builtin">sum</span>),
              size=annotations.groupby(<span class="org-string">'chip_id'</span>)[<span class="org-string">'mol_hs'</span>].agg(np.<span class="org-builtin">sum</span>), log2=<span class="org-constant">True</span>)

<span class="org-variable-name">models</span> = []
<span class="org-keyword">for</span> k <span class="org-keyword">in</span> log_cpm:
  <span class="org-keyword">if</span> k <span class="org-keyword">in</span> bulk_log_tpm:
    <span class="org-variable-name">y</span>, <span class="org-variable-name">x</span> = log_cpm[k].align(bulk_log_tpm[k], join=<span class="org-string">'inner'</span>)
    <span class="org-variable-name">m</span> = sklm.LinearRegression(fit_intercept=<span class="org-constant">True</span>).fit(x.values.reshape(-1, 1), y)
    models.append({<span class="org-string">'ind'</span>: k, <span class="org-string">'coef_'</span>: m.coef_[0], <span class="org-string">'intercept_'</span>: m.intercept_, <span class="org-string">'score'</span>: m.score(x.values.reshape(-1, 1), y)})
<span class="org-variable-name">models</span> = pd.DataFrame.from_dict(models).set_index(<span class="org-string">'ind'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">models.sort_values(<span class="org-string">'score'</span>, ascending=<span class="org-constant">False</span>)
</pre>
</div>

<pre class="example">
coef_  intercept_     score
ind
NA19209  0.732135    0.621440  0.708942
NA18870  0.723933    0.639626  0.702726
NA19101  0.729250    0.821881  0.694528
NA18511  0.707770    0.595565  0.693847
NA18912  0.688394    0.829654  0.692485
NA18520  0.731075    0.540232  0.685123
NA18502  0.712065    0.709610  0.681293
NA19127  0.713962    0.681134  0.680976
NA18856  0.734898    0.524328  0.680601
NA19225  0.679871    0.987852  0.680592
NA19116  0.713121    0.788535  0.679035
NA18505  0.684536    0.776815  0.677933
NA18862  0.696277    0.826739  0.677919
NA19190  0.712649    0.631736  0.677498
NA18859  0.701953    0.727513  0.677474
NA18858  0.725889    0.689016  0.674898
NA19204  0.713031    0.618002  0.673072
NA19128  0.713168    0.659729  0.672981
NA19210  0.705317    0.620322  0.671795
NA18852  0.725955    0.547926  0.671536
NA18489  0.698290    0.886555  0.670849
NA19160  0.717357    0.639725  0.670532
NA18853  0.679031    0.844840  0.668854
NA19207  0.724029    0.594377  0.667087
NA19102  0.720588    0.552334  0.665105
NA18501  0.686617    0.784491  0.664648
NA19140  0.685028    0.794710  0.664363
NA19257  0.668304    0.932599  0.664279
NA19143  0.711429    0.596842  0.661830
NA19152  0.696553    0.801533  0.659410
NA19099  0.721284    0.652934  0.658266
NA19153  0.708352    0.790061  0.657488
NA19108  0.691788    0.722956  0.654361
NA18913  0.713234    0.619717  0.652851
NA18517  0.671620    0.830685  0.649466
NA19093  0.720266    0.673872  0.646884
NA18519  0.715715    0.727332  0.645861
NA19206  0.701227    0.721929  0.644383
NA19114  0.671593    1.001002  0.643536
NA18855  0.708293    0.735871  0.642197
NA19130  0.699075    0.643755  0.641843
NA19098  0.696567    0.798366  0.641485
NA19193  0.704863    0.794734  0.640733
NA18907  0.715533    0.630487  0.639015
NA18508  0.659032    0.924178  0.631646
NA18522  0.686510    0.881811  0.631580
NA19144  0.663290    1.017565  0.631238
NA18507  0.641305    1.306322  0.630363
NA19159  0.712498    0.504794  0.628940
NA18498  0.695710    0.710972  0.624978
</pre>
</div>
</div>

<div id="outline-container-org50c3440" class="outline-2">
<h2 id="org50c3440">Plot bulk vs. pooled subsets</h2>
<div class="outline-text-2" id="text-org50c3440">
<p>
Plot concordance between bulk vs pools of single cells, focusing on genes
which have log-transformed expression at least 1 in both assays.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">plot_concordance_by_num_cells</span>(individual, umi, annotations, bulk_tpm, output_dir):
  <span class="org-variable-name">bulk_tpm</span> = bulk_tpm[individual].to_frame()
  <span class="org-variable-name">umi</span> = umi.loc[:,(annotations[<span class="org-string">'chip_id'</span>] == individual).values]
  <span class="org-keyword">for</span> num_cells <span class="org-keyword">in</span> [1, 10, 50, 100, 200]:
    <span class="org-variable-name">pooled_cpm</span> = cpm(umi.sample(n=num_cells, axis=1).<span class="org-builtin">sum</span>(axis=1).to_frame(), log2=<span class="org-constant">True</span>)
    plot_concordance(
      x=pooled_cpm,
      y=bulk_tpm,
      title=<span class="org-string">'{}, {} cell{}'</span>.<span class="org-builtin">format</span>(individual, num_cells, <span class="org-string">'s'</span> <span class="org-keyword">if</span> num_cells &gt; 1 <span class="org-keyword">else</span> <span class="org-string">''</span>),
      filename=<span class="org-string">'{}/{}-{}.svg'</span>.<span class="org-builtin">format</span>(output_dir, individual, num_cells),
      gridsize=20)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plot_concordance_by_num_cells(
  <span class="org-string">'NA18507'</span>,
  umi,
  annotations,
  bulk_log_tpm,
  <span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/analysis/figure/sc-vs-bulk.org/vs-cells/'</span>
)
</pre>
</div>

<p>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/vs-cells/NA18507-1.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/vs-cells/NA18507-10.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/vs-cells/NA18507-50.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/vs-cells/NA18507-100.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/vs-cells/NA18507-200.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>

<div id="outline-container-orgcdd234c" class="outline-2">
<h2 id="orgcdd234c">Plot pooled subsets vs. pooled subsets</h2>
<div class="outline-text-2" id="text-orgcdd234c">
<p>
Ensure that pools don't overlap by randomly sampling double the cells and
partitioning into two halves.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">plot_concordance_pooled_subsets</span>(individual, umi, annotations, output_dir):
  <span class="org-variable-name">umi</span> = umi.loc[:,(annotations[<span class="org-string">'chip_id'</span>] == individual).values]
  <span class="org-keyword">for</span> num_cells <span class="org-keyword">in</span> [1, 10, 50, 100]:
    <span class="org-variable-name">sample</span> = umi.sample(n=2 * num_cells, axis=1)
    <span class="org-variable-name">pool1</span> = cpm(sample.iloc[:,:num_cells].<span class="org-builtin">sum</span>(axis=1).to_frame(), log2=<span class="org-constant">True</span>)
    <span class="org-variable-name">pool2</span> = cpm(sample.iloc[:,num_cells:].<span class="org-builtin">sum</span>(axis=1).to_frame(), log2=<span class="org-constant">True</span>)
    plot_concordance(
      x=pool1,
      y=pool2,
      title=<span class="org-string">'{}, {} cell{}'</span>.<span class="org-builtin">format</span>(individual, num_cells, <span class="org-string">'s'</span> <span class="org-keyword">if</span> num_cells &gt; 1 <span class="org-keyword">else</span> <span class="org-string">''</span>),
      filename=<span class="org-string">'{}/{}-{}.svg'</span>.<span class="org-builtin">format</span>(output_dir, individual, num_cells),
      ylabel=<span class="org-string">'scRNA-Seq $\log_2(\mathrm{CPM} + 1)$'</span>,
      gridsize=15)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plot_concordance_pooled_subsets(
  <span class="org-string">'NA18507'</span>,
  umi,
  annotations,
  <span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/analysis/figure/sc-vs-bulk.org/subsets/'</span>
)
</pre>
</div>

<p>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/subsets/NA18507-1.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/subsets/NA18507-10.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/subsets/NA18507-50.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/subsets/NA18507-100.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>

<div id="outline-container-orgd1000d9" class="outline-2">
<h2 id="orgd1000d9">Plot bulk vs. single cell relative abundance</h2>
<div class="outline-text-2" id="text-orgd1000d9">
<p>
TPM is proportional to relative abundance.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">bulk_tpm</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/kallisto/bulk-ipsc-tpm.txt.gz'</span>, header=<span class="org-constant">None</span>, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">bulk_rho</span> = bulk_tpm.pivot(columns=0, index=1, values=2)
<span class="org-variable-name">bulk_rho.index</span> = [x.split(<span class="org-string">'.'</span>)[0] <span class="org-keyword">for</span> x <span class="org-keyword">in</span> bulk_rho.index]
<span class="org-variable-name">bulk_rho</span> = bulk_rho.<span class="org-builtin">filter</span>(items=keep_genes.index, axis=<span class="org-string">'index'</span>)
<span class="org-variable-name">bulk_log_rho</span> = np.ma.masked_invalid(np.log(bulk_rho.values)) - np.log(bulk_rho.values.<span class="org-builtin">sum</span>(axis=0))
<span class="org-variable-name">bulk_log_rho</span> = pd.DataFrame(bulk_log_rho, index=bulk_rho.index, columns=bulk_rho.columns)
</pre>
</div>

<p>
Under <a href="zinb.html">our assumed model</a>, the parameter \(\mu\) is relative abundance.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> sqlite3

<span class="org-keyword">with</span> sqlite3.connect(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/browser/browser.db'</span>) <span class="org-keyword">as</span> conn:
  <span class="org-variable-name">log_mu</span> = (pd.read_sql(
    <span class="org-string">"""select gene, ind, log_mu from params where ind in </span>
<span class="org-string">    (select chip_id from annotation group by chip_id </span>
<span class="org-string">    having count(distinct sample) &gt;= 50);"""</span>, conn)
              .pivot(index=<span class="org-string">'gene'</span>, columns=<span class="org-string">'ind'</span>, values=<span class="org-string">'log_mu'</span>))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">plot_concordance_rho</span>(bulk, sc, output_dir):
  <span class="org-variable-name">bulk</span>, <span class="org-variable-name">sc</span> = bulk.align(sc, axis=1, join=<span class="org-string">'inner'</span>)
  <span class="org-keyword">for</span> k <span class="org-keyword">in</span> bulk:
    plot_concordance(
      x=sc[k].to_frame(),
      y=bulk[k].to_frame(),
      title=k,
      xlabel=<span class="org-string">'Single cell log relative abundance'</span>,
      ylabel=<span class="org-string">'Bulk log relative abundance'</span>,
      filename=<span class="org-string">'{}/{}.svg'</span>.<span class="org-builtin">format</span>(output_dir, k))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plot_concordance_rho(
  bulk_log_rho,
  log_mu,
  <span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/analysis/figure/sc-vs-bulk.org/vs-sc-mean'</span>)
</pre>
</div>

<p>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/vs-sc-mean/NA18501.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/vs-sc-mean/NA18502.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
<object type="image/svg+xml" data="figure/sc-vs-bulk.org/vs-sc-mean/NA18505.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-03-28 Wed 20:24</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
