---
title: "Introduction"
author: "Po-Yuan Tung"
date: 2018-01-25
output: html_document
---

<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
```{r read-chunk, include=FALSE, cache=FALSE}
knitr::read_chunk("chunks.R")
```

<!-- Update knitr chunk options -->
```{r knitr-opts-chunk, include=FALSE}
```

<!-- Insert the date the file was last updated -->
```{r last-updated, echo=FALSE, results='asis'}
```

<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
```{r code-version, echo=FALSE, results='asis'}
```

The importance of robustness and the regulation of noise as mechanisms that maintain high evolutionary fitness is evident[@Kirschner1998;@Kitano2004;@Masel2009;@Wolf2015].
Yet, we still have a relatively poor understanding of how robustness is achieved and how expression noise is being regulated at the molecular level.
In model organisms, robustness and evolvability can be studied using experimental evolution approaches.
Relative robustness is typically quantified with respect to the change in variation of a specific trait when the experimental perturbation is applied [@Barrick2013; @deVisser2003].
In such experiments, the phenotypic outcomes rather than the underlying mechanisms of robustness are measured.
To take first steps towards understanding how robustness is regulated in humans, we proposed to focus on a molecular phenotype, namely gene expression levels.

With few exceptions [@Deng2014;@Marinov2014;@Gasch2017], experimental evolution studies have always considered population-average measurements of phenotypes using entire organisms, tissues, or cell cultures.
However, as was often suggested in the literature, to truly understand how robustness is established and encoded in the genome, one needs to consider variation in phenotypes across individual cells[@Raser2004;@Kaern2005;@Balazsi2011;@Holmes2017].
The motivation to learn more about the property of robustness has led to the advent of new approaches to study ‘noise genetics’ – essentially using existing paradigms to characterize naturally occurring cell-to-cell variation instead of the population-average response to an artificial perturbation[@Farkash-Amar2014].
The idea of using the probability of gene expression at the single cell level as a genetic phenotype, termed ‘particle genetics’, has also been introduced to study complex traits [@Yvert2014].
While a large number of studies have focused on the robustness of particular cellular functions or individual genes in humans [@Wills2013;@Shaffer2014;@Halpern2015;@Faure2017;@Hendy2017], genome-wide analysis and mapping of the genetic loci underlying robustness has yet to be performed.

Recent advances in single-cell and single-molecule techniques have revealed that gene expression noise caused by stochastic fluctuations is a complex genetic trait in yeasts[@Ansel2008].
As a result, research is also aimed at understanding how expression noise is controlled and interpreted within gene regulatory networks [@Eldar2010;@Chalancon2012].
However, a lack of appropriate quantitative experimental approaches has made the mapping of genetic loci responsible for differential cell-cell variation (or gene expression noise) extremely difficult.
One methodology that has been proved powerful as a tool to dissect the genetic architecture of complex phenotypes is the quantitative trait loci (QTL) mapping approach.
In the context of molecular phenotypes, such as gene expression levels and epigenetic regulatory features, QTL studies in a wide range of species, including humans, have provided important insights into the genetic basis for inter-individual variation in gene regulation [@Gilad2008; @Pickrell2010; @Majewski2011].
Since gene expression levels of population-average measurements are proven to be genetically controlled, it stands to reason that the levels of acceptable regulatory noise (or the degree of robustness) would also be genetically encoded.
Indeed, the identification of QTLs influencing stochastic cell–cell variability in yeast lends some measure of support to this notion[@Chuffart2016].
Yet, regulatory QTL studies in mammalian systems have not yet considered this aspect.
All studies to date focused on the mean expression levels across millions of cells, lacking the expression profiling information at the single cell level.

Induced pluripotency stem cells (iPSCs)[@Takahashi2006] are a tantalizingly flexible system with which many different developmental and cell lineage determination pathways can be explored.
Recent studies have demonstrated that iPSCs can be utilized as a powerful tool for studies of the genetic association with variation in gene regulation[@Thomas2015; @Burrows2016; @DeBoever2017; @Kilpinen2017].
In particular, iPSCs provide a means by which to examine the contribution of changes in gene regulation to human evolution and diversity[@Romero2012; @Banovich2018].
Importantly, the heterogeneous gene expression levels between individual cells observed in population of iPSCs has been proposed to reflect the alternate cell states that represent multiple differentiation potentials[@Narsinh2011].
To understand the underlying mechanisms of such plasticity in human iPSCs, quantification of gene expression at the single-cell level is required.
Nevertheless, the lack of an iPSC system for single cell studies of robustness has hindered the investigation of regulatory gene expression noise during differentiation and development in humans.

In this genome-wide study, the relative genetic robustness is measured as the inter-individual difference in magnitude of variation in gene expression levels across individual cells.
To do so, we applied a QTL approach to map inter-individual variation in gene expression noise in a panel of iPSCs reprogrammed from deeply characterized Yoruba HapMap lymphoblastoid cell lines (LCLs)[@Banovich2018]. 
Using the Fluidigm C1 system, we collected single-cell RNA-seq from 54 individuals using the technique of unique molecular identifiers (UMIs).
This approach allows for the detection of mRNA molecule counts and hence provides more accurate estimation of the mean and the variance of expression level for each gene.
We analyzed the variability in gene regulation across individual cells as a measure of expression regulatory noise.
In addition, we developed a novel QTL mapping framework specifically applied to estimates of variance rather than average.
To our knowledge, this is the first attempt to map genome wide robustness QTLs in humans.
